storage/zynq7000-flash/qspi.c

```diff
@@ -0,0 +1,436 @@
+	*(qspi_common.base + cr) &= ~(1 << 26);
+	/* Set FIFO width 32 bits */
+	*(qspi_common.base + cr) |= (0x3 << 6);
+	/* Set clock phase and polarity */
+	*(qspi_common.base + cr) &= ~(0x3 << 1);
+
+
+	/* Enable manual mode and manual CS */
+	*(qspi_common.base + cr) |= (0x3 << 14);
+
+	/* Disable linear mode */
+	*(qspi_common.base + lqspi_cr) = 0;
+	qspi_dataMemoryBarrier();
+}
+
+
+static int qspi_setAmbaClk(int dev, unsigned int state)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_ambaclock;
+
+	ctl.ambaclock.dev = dev;
+	ctl.ambaclock.state = state;
+
+	return platformctl(&ctl);
+}
+
+
+static int qspi_initClk(void)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+
+	ctl.devclock.dev = pctl_ctrl_lqspi_clk;
+	ctl.devclock.clkact0 = 0x1;
+	ctl.devclock.clkact1 = 0x1;
+	ctl.devclock.srcsel = 0;
+	ctl.devclock.divisor0 = 5;
+
+	return platformctl(&ctl);
+}
+
+
+static int qspi_setPin(uint32_t pin)
+{
+	platformctl_t ctl;
+
+	if (pin < pctl_mio_pin_01 && pin > pctl_mio_pin_06)
+		return -EINVAL;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_mio;
+
+	ctl.mio.pin = pin;
+	ctl.mio.l0 = 0x1;
+	ctl.mio.l1 = ctl.mio.l2 = ctl.mio.l3 = 0;
+	ctl.mio.pullup = 0;
+	ctl.mio.speed = 0x1;
+	ctl.mio.ioType = 0x1;
+	ctl.mio.disableRcvr = 0;
+	ctl.mio.triEnable = 0;
+
+	if (pin == pctl_mio_pin_01) {
+		ctl.mio.pullup = 1;
+		ctl.mio.speed = 0x0;
+	}
+
+	return platformctl(&ctl);
+}
+
+
+static int qspi_initPins(void)
+{
+	int res, i;
+	static const int pins[] = {
+		pctl_mio_pin_01, /* Chip Select */
+		pctl_mio_pin_02, /* I/O */
+		pctl_mio_pin_03, /* I/O */
+		pctl_mio_pin_04, /* I/O */
+		pctl_mio_pin_05, /* I/O */
+		pctl_mio_pin_06  /* CLK */
+	};
+
+	for (i = 0; i < sizeof(pins) / sizeof(pins[0]); ++i) {
+		res = qspi_setPin(pins[i]);
+		if (res < 0)
+			break;
+	}
+
+	return res;
+}
+
+
+int qspi_deinit(void)
+{
+	int res;
+
+	qspi_stop();
+
+	res = qspi_setAmbaClk(pctl_amba_lqspi_clk, 0);
+	if (res < 0)
+		return res;
+
+	return res;
+}
+
+
+int qspi_init(void)
+{
+	int res;
+
+	/* Set IO PLL as source clock and set divider:
+	 * IO_PLL / 0x5 :  1000 MHz / 5 = 200 MHz     */
+	res = qspi_initClk();
+	if (res < 0)
+		return -EIO;
+
+	/* Enable clock */
+	res = qspi_setAmbaClk(pctl_amba_lqspi_clk, 1);
+	if (res < 0)
+		return -EIO;
+
+	/* Single SS, 4-bit I/O */
+	res = qspi_initPins();
+	if (res < 0)
+		return -EIO;
+
+	qspi_common.base = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, 0xe000d000);
+	if (qspi_common.base == MAP_FAILED)
+		return -ENOMEM;
+
+	res = condCreate(&qspi_common.cond);
+	if (res < 0) {
+		munmap((void *)qspi_common.base, _PAGE_SIZE);
+		qspi_deinit();
+		return -ENOENT;
+	}
+
+	res = mutexCreate(&qspi_common.lock);
+	if (res < 0) {
+		munmap((void *)qspi_common.base, _PAGE_SIZE);
+		resourceDestroy(qspi_common.cond);
+		qspi_deinit();
+		return -ENOENT;
+	}
+
+	interrupt(51, qspi_irqHandler, NULL, qspi_common.cond, &qspi_common.inth);
```

Comment:
irq numbers should be in the platfrom header in the kernel include - maybe it's a good time to add them there

Reply:
The new task was created to add irq numbers to the platform header. I think that this PR can be merged without it and will be updated in the future.

Reply:
Ok